Logging: library module to log over multiple loggers.

asheshambasta opened the pull request:
Commits: 

- Add a logging module for multiple-logger logging.
- [Minor] Fix warnings (unused imports)

Adds logging helpers to log over multiple loggers, each with their own configurations. 
Support for LogCallback is present, but cannot be parsed via the CLI. @thusc, the recommended approach here can be that we have a preset list of callback loggers and they can be parsed as enums on the cli. We should see how this evolves in the future, if there is a need for that, however.

asheshambasta committed:
Add a logging module for multiple-logger logging.

The new logging module, as we discussed, @thusc, provides
functionality to log over multiple loggers.

This, is, from what I can gather, fully supportive of our
requirements.

Additionally, I've kept convenience in mind in keeping with having a
typeclass that we needn't care about when dealing with logging in some `m`.

asheshambasta committed:
[Minor] Fix warnings (unused imports)

asheshambasta commented:
@noteed would you mind taking a look at this?

noteed commented on src/MultiLogging.hs:
```diff
@@ -0,0 +1,173 @@
+data LoggingConf = LoggingConf
+  { _lcOutputs     :: [FL.LogType] -- ^ Multiple logging outputs: logging is disabled if this list is empty.
+  , _lcRootAppName :: L.AppName  -- ^ Application name to use at the root of the logger. 
+  , _lcBufSize     :: FL.BufSize -- ^ Buffer size to observe for logging.
```
This seems unused, so I'm not yet sure what this is used for.

noteed commented on src/Logging.hs:
```diff
@@ -51,18 +51,20 @@ module Logging
+import qualified Options.Applicative           as A
+import "protolude" Protolude
 
 -- | A type alias for convenience
```
What do you think about adding something like "A Control.Monad.Log Logger where the environment is an AppName." ?

noteed commented on src/MultiLogging.hs:
```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE ApplicativeDo #-}
```
What do you think of adding something like: "This module exposes functions to help log to multiple FastLogger loggers, using an API similar to monad-log (which is itself also used)."

noteed reviewed (approved):
I've added some comments above, otherwise LGTM.

asheshambasta committed:
[PR feedback] Add docs

@thusc @noteed for https://github.com/noteed/start-servant/pull/18/files#r825308405

asheshambasta committed:
[PR feedback] Add docs, remove lcBufSize

@thusc @noteed for:
- https://github.com/noteed/start-servant/pull/18/files#r825309437
- https://github.com/noteed/start-servant/pull/18/files#r825306841

asheshambasta commented:
@noteed re-requested review.

noteed reviewed (approved):
:shipit:
Thanks !
